### PR TITLE
Adds Python 3.3, 3.4 and 3.5 to supported versions

### LINKED ISSUE #7: Python 3 support
Python 3 support using the six module is in progress. Finish the port for release 3.2.1.

### LINKED ISSUE #8: pydap.responses.das attribute order is non-deterministic
## Description

I have been using `tox` to test the develop branch across Python versions 2.6, 2.7 and 3.3. I have noticed that when running the tests with `tox`, I get sporadic failures (ranging between 0 and 4 failures per run) in a number of tests, most commonly in test_responses_das.py.

When running the tests with the standard `nosetests` command, [all of the tests pass](https://travis-ci.org/pacificclimate/pydap/builds/25041886).

I think that I've narrowed it down to the fact that `tox` [randomizes the Python hash seed](http://testrun.org/tox/latest/example/basic.html#special-handling-of-pythonhashseed) on every run, so (correctly) the order of dictionary items is not guaranteed unless you take measures to do the ordering yourself.
## Example

Here is an example of running the test suite with different hash seeds:

```
james@basalt ~/code/git/pydap $ tox -e py26 --hashseed 4016568519 
py26 develop-inst-nodeps: /home/james/code/git/pydap
py26 runtests: PYTHONHASHSEED='4016568519'
py26 runtests: commands[0] | pip install -e .[testing,functions,tests]
Obtaining file:///home/james/code/git/pydap
  Running setup.py (path:/home/james/code/git/pydap/setup.py) egg_info for package from file:///home/james/code/git/pydap
  Installing extra requirements: 'testing,functions,tests'
Requirement already satisfied (use --upgrade to upgrade): Numpy in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): requests in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): singledispatch in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): Webob in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): Jinja2 in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): docopt in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): gunicorn in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): six>=1.4.0 in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): ordereddict in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): gsw in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): coards in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): scipy in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): WebTest in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): beautifulsoup4 in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): nose in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): coverage in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): markupsafe in ./.tox/py26/lib/python2.6/site-packages (from Jinja2->Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): waitress>=0.8.5 in ./.tox/py26/lib/python2.6/site-packages (from WebTest->Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): setuptools in ./.tox/py26/lib/python2.6/site-packages (from waitress>=0.8.5->WebTest->Pydap==3.2)
Installing collected packages: Pydap
  Running setup.py develop for Pydap
    Creating /home/james/code/git/pydap/.tox/py26/lib/python2.6/site-packages/Pydap.egg-link (link to src)
    Pydap 3.2 is already the active version in easy-install.pth
    Installing dods script to /home/james/code/git/pydap/.tox/py26/bin
    Installing pydap script to /home/james/code/git/pydap/.tox/py26/bin

    Installed /home/james/code/git/pydap/src
Successfully installed Pydap
Cleaning up...
py26 runtests: commands[1] | python setup.py nosetests --pdb
running nosetests
running egg_info
writing requirements to src/Pydap.egg-info/requires.txt
writing namespace_packages to src/Pydap.egg-info/namespace_packages.txt
writing src/Pydap.egg-info/PKG-INFO
writing top-level names to src/Pydap.egg-info/top_level.txt
writing dependency_links to src/Pydap.egg-info/dependency_links.txt
writing entry points to src/Pydap.egg-info/entry_points.txt
reading manifest file 'src/Pydap.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'src/Pydap.egg-info/SOURCES.txt'
.....SS......................./home/james/code/git/pydap/src/pydap/handlers/dap.py:145: DeprecationWarning: converting an array with ndim > 0 to an index will result in an error in the future
  out.append(data[:n])
/home/james/code/git/pydap/src/pydap/handlers/dap.py:146: DeprecationWarning: converting an array with ndim > 0 to an index will result in an error in the future
  data = data[n + (-n % 4):]
...
......................................................................................................................................................................................Ran 311 tests in 3.132s

OK (SKIP=2)
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
  py26: commands succeeded
  congratulations :)
```

Compare this to a separate run with a different hash seed:

```
james@basalt ~/code/git/pydap $ tox -e py26 --hashseed 3775216848
py26 develop-inst-nodeps: /home/james/code/git/pydap
py26 runtests: PYTHONHASHSEED='3775216848'
py26 runtests: commands[0] | pip install -e .[testing,functions,tests]
Obtaining file:///home/james/code/git/pydap
  Running setup.py (path:/home/james/code/git/pydap/setup.py) egg_info for package from file:///home/james/code/git/pydap
  Installing extra requirements: 'testing,functions,tests'
Requirement already satisfied (use --upgrade to upgrade): Numpy in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): requests in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): singledispatch in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): Webob in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): Jinja2 in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): docopt in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): gunicorn in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): six>=1.4.0 in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): ordereddict in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): gsw in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): coards in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): scipy in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): WebTest in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): beautifulsoup4 in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): nose in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): coverage in ./.tox/py26/lib/python2.6/site-packages (from Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): markupsafe in ./.tox/py26/lib/python2.6/site-packages (from Jinja2->Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): waitress>=0.8.5 in ./.tox/py26/lib/python2.6/site-packages (from WebTest->Pydap==3.2)
Requirement already satisfied (use --upgrade to upgrade): setuptools in ./.tox/py26/lib/python2.6/site-packages (from waitress>=0.8.5->WebTest->Pydap==3.2)
Installing collected packages: Pydap
  Running setup.py develop for Pydap
    Creating /home/james/code/git/pydap/.tox/py26/lib/python2.6/site-packages/Pydap.egg-link (link to src)
    Pydap 3.2 is already the active version in easy-install.pth
    Installing pydap script to /home/james/code/git/pydap/.tox/py26/bin
    Installing dods script to /home/james/code/git/pydap/.tox/py26/bin

    Installed /home/james/code/git/pydap/src
Successfully installed Pydap
Cleaning up...
py26 runtests: commands[1] | python setup.py nosetests
running nosetests
running egg_info
writing dependency_links to src/Pydap.egg-info/dependency_links.txt
writing requirements to src/Pydap.egg-info/requires.txt
writing namespace_packages to src/Pydap.egg-info/namespace_packages.txt
writing top-level names to src/Pydap.egg-info/top_level.txt
writing src/Pydap.egg-info/PKG-INFO
writing entry points to src/Pydap.egg-info/entry_points.txt
reading manifest file 'src/Pydap.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'src/Pydap.egg-info/SOURCES.txt'
.....SS......................./home/james/code/git/pydap/src/pydap/handlers/dap.py:145: DeprecationWarning: converting an array with ndim > 0 to an index will result in an error in the future
  out.append(data[:n])
/home/james/code/git/pydap/src/pydap/handlers/dap.py:146: DeprecationWarning: converting an array with ndim > 0 to an index will result in an error in the future
  data = data[n + (-n % 4):]
......................................................................................................................................................................................FF.....F.................................F.........................................................
======================================================================
FAIL: Test the generated DAS response.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/james/code/git/pydap/src/pydap/tests/test_responses_das.py", line 113, in test_body
    """)
AssertionError: u'Attributes {\n    String description "A simple grid for testing.";\n    SimpleGrid {\n    }\n    x {\n        String axis "X";\n        String units "degrees_east";\n    }\n    y {\n        String axis "Y";\n        String units "degrees_north";\n    }\n}\n' != 'Attributes {\n    String description "A simple grid for testing.";\n    SimpleGrid {\n    }\n    x {\n        String units "degrees_east";\n        String axis "X";\n    }\n    y {\n        String units "degrees_north";\n        String axis "Y";\n    }\n}\n'

======================================================================
FAIL: Test the generated DAS response.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/james/code/git/pydap/src/pydap/tests/test_responses_das.py", line 88, in test_body
    """)
AssertionError: u'Attributes {\n    nested {\n        Int32 value 42;\n    }\n    String description "A simple sequence for testing.";\n    cast {\n        id {\n        }\n        lon {\n            String axis "X";\n        }\n        lat {\n            String axis "Y";\n        }\n        depth {\n            String axis "Z";\n        }\n        time {\n            String axis "T";\n            String units "days since 1970-01-01";\n        }\n        temperature {\n        }\n        salinity {\n        }\n        pressure {\n        }\n    }\n}\n' != 'Attributes {\n    String description "A simple sequence for testing.";\n    nested {\n        Int32 value 42;\n    }\n    cast {\n        id {\n        }\n        lon {\n            String axis "X";\n        }\n        lat {\n            String axis "Y";\n        }\n        depth {\n            String axis "Z";\n        }\n        time {\n            String units "days since 1970-01-01";\n            String axis "T";\n        }\n        temperature {\n        }\n        salinity {\n        }\n        pressure {\n        }\n    }\n}\n'

======================================================================
FAIL: Test the generated DAS response.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/james/code/git/pydap/src/pydap/tests/test_responses_das.py", line 154, in test_body
    """)
AssertionError: u'Attributes {\n    types {\n        nested {\n            Float64 float 1000;\n            Int32 array 1;\n            String string "bar";\n            Int32 list 42, 43;\n        }\n        String key "value";\n        b {\n        }\n        i32 {\n        }\n        ui32 {\n        }\n        i16 {\n        }\n        ui16 {\n        }\n        f32 {\n        }\n        f64 {\n        }\n        s {\n        }\n        u {\n        }\n    }\n}\n' != 'Attributes {\n    types {\n        String key "value";\n        nested {\n            Int32 array 1;\n            Float64 float 1000;\n            Int32 list 42, 43;\n            String string "bar";\n        }\n        b {\n        }\n        i32 {\n        }\n        ui32 {\n        }\n        i16 {\n        }\n        ui16 {\n        }\n        f32 {\n        }\n        f64 {\n        }\n        s {\n        }\n        u {\n        }\n    }\n}\n'

======================================================================
FAIL: Test that the data redirect works with a subset request.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/james/code/git/pydap/src/pydap/tests/test_responses_html.py", line 88, in test_post_multiple_selection
    res.location, "http://localhost/.ascii?sequence.int,sequence.byte")
AssertionError: 'http://localhost/.ascii?sequence.byte,sequence.int' != 'http://localhost/.ascii?sequence.int,sequence.byte'

Ran 311 tests in 3.127s

FAILED (SKIP=2, failures=4)
ERROR: InvocationError: '/home/james/code/git/pydap/.tox/py26/bin/python setup.py nosetests'
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
ERROR:   py26: commands failed
```

If you look very closely there, you'll notice that the attribute "String description ..." is transposed with the attribute "nested {...}". The behaviour is similar for Python 2.7.
## How to fix

This obvious fix seems to be to make stronger guarantees about the ordering of attributes. I don't necessarily know what those guarantees _should_ be, but there are a couple options:
1. Use and an `OrderedDict` for the `DapType.attributes` dictionary. In this case attributes returned in the DAS response will be ordered according to insertion order.
2. Sort the dict keys before iterating over them in `pydap.responses.das:das()`. There appear to be three places where this would need to be changed. In this case attributes returned in the DAS response will be ordered alphabetically.

I can fix it in either way, if someone would like to provide an opinion on the better option (or comment otherwise).

### PR DESCRIPTION
**Pull request by jameshiebert**
### Summary

This PR ports the remainder of the Pydap code to run on Python version 2.6, 2.7 and 3.3. It closes #7 and fixes #8.

Much of this exercise was straightforward, but there were a few points issues on which feedback from others would be valued.
### OPeNDAP, Python3, bytes, Unicode and ASCII

In Python 3, [one needs to be very specific](http://www.slideshare.net/dabeaz/mastering-python-3-io-version-2) about whether one is using raw bytes vs. text in encoded with unicode vs. text encoded with some other encoding, etc. These are different base types in different Python versions, so I've used the `six` module's `text_type` where appropriate.

The problems is that a careful reading of the [DAP 2.0 spec](http://www.opendap.org/pdf/ESE-RFC-004v1.1.pdf) provides little clarity about what the text encoding of a DAP response is _supposed_ to be. Actually, section 3.2.3 does specify that the String Type is limited to 7-bin ASCII characters and section 5 places further restrictions on the character sets of variable names.

Because of this, I have changed numerous places in the code--particularly in the DAS and DDS handlers/responses--setting the output encoding to ASCII rather than UTF-8. I don't actually like this... it's limiting the internationalization support and also a wide variety of scientific symbols that would be used in DAP attributes. However, it's true to the DAP spec by my reading. Feedback on this is welcome.
### Conflicting HTTP Request Frameworks

Most of the Pydap client/server code uses `webob`'s `Request` and `Response` objects. However, most (all?) of the testing code uses the `requests` package and the associated API. Both of these packages are fantastic in their own right, but they have slightly different API's.

The test suite [_does_ pass](https://travis-ci.org/pacificclimate/pydap/builds/25826934) under all three version of Python, however, I'm nervous that the testing code isn't actually testing the production code (particularly w.r.t. content iteration). In this context, I would just as soon use `webob` across the board.
### Monkeypatching sys.stdin is a bad idea

I [removed one problematic test](https://github.com/pacificclimate/pydap/commit/afab06949e0a54cc38146e3f5e0fa4498612b192) because it was failing, and the setup function monkeypatched sys.stdin, which made it particularly heinous to debug (this makes `nosetests --pdb` disfunctional).

That test just tested a function that dumps to stdout, so I just marked it with a no cover pragma. If there's a better solution, I'm open to it.

### CODE REVIEW

**Comment by shoyer** [comment]
I would like to be able to use pydap on Python 3 (at least as a client).

What is this PR waiting on?

**Comment by jameshiebert** [comment]
In my opinion, this PR changes things significantly enough (as discussed above), that I would want some code review on it before seeing it merged. That could be from the package author or otherwise, I suppose.

**Comment by shoyer** [comment]
Well I don't know pydap internals but I'm happy to take a closer look here. My general view is that any Python 3 support is better than nothing. We could even consider Python 3 support an officially "experimental" feature for now if that helps. In particular, if we end up regretting decoding all attributes as ASCII that will be easy enough to fix later.

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,25 @@
...
-script: nosetests
 branches:
-  only:
-      - develop
+    only:
+        - develop
+        - 2to3
+
+before_install:
+    # In python 2.7 we have scipy/numpy through the Ubuntu package manager
```
Comment (shoyer) [inline .travis.yml]: If this works, great! Another option is to install scipy/cython/numpy with conda, which will be much faster. For an example, see here:
https://github.com/xray/xray/blob/master/.travis.yml

File: setup.cfg [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [nosetests]
 match=^test
-where=src/pydap
+#where=src/pydap
```
Comment (shoyer) [inline setup.cfg]: probably want to delete this rather than commenting out?

File: setup.py [hunk 3]
```diff
@@ -12,8 +12,8 @@
     'Jinja2',
     'docopt',
     'gunicorn',
-    'PasteDeploy',
-    'six',
+#    'PasteDeploy',
```
Comment (shoyer) [inline setup.py]: same as above

File: .travis.yml [hunk 4]
```diff
@@ -1,10 +1,25 @@
...
+          sudo apt-get install -qq gfortran libblas-dev liblapack-dev cython;
+      fi
+
+install:
+    - pip install -e .[testing,functions,tests]
+
+python:
+    - 2.6
+    - 2.7_with_system_site_packages
+    - 3.3
```
Comment (shoyer) [inline .travis.yml]: It would be nice to add 3.4 to the build matrix, too

**Comment by shoyer** [comment]
As for unicode/bytes: from the DAP spec, it sure does look like DAP was only envisioned to support ASCII.

That said...

1. We are fortunate that ASCII is a proper subset of UTF-8
2. I find it difficult to imagine that any future DAP spec that *did* allow text encodings beyond ASCII would choose anything other than UTF-8.

So I think it would be totally safe to *decode* strings as UTF-8 in the client library (rather than ASCII). That said, given that no DAP servers are serving UTF-8, I don't know if there's any point.

One other consideration (and I'm not sure if you've addressed this or not) is what to do with string arrays. Given that NumPy still lacks support for variable width unicode (UTF-8), we might want to keep these arrays encoded as bytes (the numpy `S` dtype) and let users decode them. The other choices are fixed width unicode or python `str` objects in a `dtype=object` array (what pandas does).

Two points of context for what netCDF libraries do on Python 3 with bytes/character types (e.g., from netCDF3 files):
- The netCDF4-Python library decodes attributes as strings and but keeps arrays as bytes.
- scipy.io keeps everything (attributes and arrays) as bytes.

So I guess I'm suggesting copying the behavior of netCDF4-python.

**Comment by shoyer** [comment]
For more context on string arrays in Python 3, read the numpy mailing list discussion:
http://mail.scipy.org/pipermail/numpy-discussion/2015-February/072311.html

Given that data in pydap is actually transmitted over a wire, decoding to fixed-width unicode (numpy's `U` dtype) could make sense. But I still think it's better to remain consistent with other scientific Python libraries.

File: src/pydap/client.py [hunk 5]
```diff
@@ -70,10 +70,22 @@ def open_file(dods, das=None):
...
+    with open(dods, "rt", buffering=1, encoding='ascii', newline='\n', errors='ignore') as f:
+        for line in f:
+            if line.strip() == 'Data:':
+                break
+            dds += line
+    dataset = build_dataset(dds)
+    pos = len(dds) + len('Data:\n')
+
     with open(dods, "rb") as f:
-        dds, data = f.read().split(b'\nData:\n', 1)
```
Comment (QuLogic) [inline src/pydap/client.py]: Why doesn't this work?

File: src/pydap/client.py [hunk 6]
```diff
@@ -70,10 +70,22 @@ def open_file(dods, das=None):
...
-        dataset = build_dataset(dds)
-        dataset.data = unpack_data(data, dataset)
+        f.seek(pos)
+        dataset.data = unpack_data(f, dataset)
 
     if das is not None:
         with open(das) as f:
```
Comment (QuLogic) [inline src/pydap/client.py]: No encoding for this one?

File: src/pydap/handlers/dap.py [hunk 7]
```diff
@@ -46,12 +49,12 @@ def __init__(self, url):
         ddsurl = urlunsplit((scheme, netloc, path + '.dds', query, fragment))
         r = requests.get(ddsurl)
         r.raise_for_status()
-        dds = r.text.encode('utf-8')
+        dds = r.text.decode('ascii')
```
Comment (QuLogic) [inline src/pydap/handlers/dap.py]: Isn't this the result from `requests.get` (of _the_ requests module)? `r.text` should already be decoded.

File: src/pydap/handlers/dap.py [hunk 8]
```diff
@@ -258,15 +262,24 @@ def __iter__(self):
...
+        # Fast forward past the DDS header
+        # the pattern could span chunk boundaries though so make sure to check
+        i = r.iter_content(1024)
+        previous_chunk = b''
+        this_chunk = b''
+        pattern = b'Data:\n'
+        while not re.search(pattern, previous_chunk + this_chunk):
+            previous_chunk = this_chunk
+            this_chunk = next(i)
+        end_chunk = (previous_chunk + this_chunk).split(b'Data:\n', 1)[1]
```
Comment (QuLogic) [inline src/pydap/handlers/dap.py]: If you save the result of `re.search`, you can get the index on which to split directly with `m.end()`.

File: src/pydap/handlers/dap.py (hunk 7 above)
Comment (jameshiebert) [inline src/pydap/handlers/dap.py]: Yeah, I think you're right. This is one of those cases that I discussed above where the code proper and the testing code don't use the same HTTP frameworks. So... the tests pass, but it's not really testing the right thing. Looking into a better solution to this...

File: src/pydap/handlers/dap.py (hunk 8 above)
Comment (jameshiebert) [inline src/pydap/handlers/dap.py]: Good call. Done.

**Comment by lewismc** [comment]
Will pull and check today.

**Comment by lewismc** [comment]
Right now there is an issue with the paver commands. I am not able to install the codebase. Is the idea to move away from Paver in this PR?

**Comment by lewismc** [comment]
@jameshiebert @shoyer @QuLogic PING, folks any idea how I install the codebase from this branch? I've tried to use the Paver command and I am getting nowhere. I think it is really important to have Pydap maintained as it is a good library. This pull request would really help keep the codebase alive.
I am not able to test the codebase unless I know whether Paver is the correct build system or not. Regardless, can someone please tell me how to install using Paver? Thanks in advance.

**Comment by shoyer** [comment]
RE paver: I have never used paver before, but it looks like it doesn't support Python 3.5 yet (https://github.com/paver/paver/issues/162). I so agree, let's drop paver is favor of standard Python build tools (e.g., a `setup.py` file). It's much better to use standard tools that people are familiar with.

To that end, maybe @jameshiebert can explain why he replaced requests with webob in his latest commits in  this PR?

I'm happy to provide input and code review, but I'm not in a position to finish up this PR or maintain pydap. I have too many other open source commitments, and don't even use pydap anymore!

@lewismc if you are interested and able to help, perhaps you could help @jameshiebert finish this up and issue a release?

**Comment by lewismc** [comment]
OK, I'll push a setup.py right now to @jameshiebert account to add to this PR. This will atleast enable me to test it out locally. I would be very happy to help get a release working. We use Pydap in a number of projects and have Python 3.X support would be excellent.

**Comment by QuLogic** [comment]
Yes, sorry, like @shoyer, I can really only provide some review and perhaps assistance with Python 3 related issues, as I don't really use pydap at the moment.

**Comment by jameshiebert** [comment]
Hey all. Sorry for the radio silence... lots on the go over here.

Re Paver: yeah, I never understood why @robertodealmeida used paver over the regular Python build tools, so I'm happy to drop support for that in the next release.

> can explain why he replaced requests with webob in his latest commits in this PR?

I just wanted to simplify... rather than using two different libraries that work towards the same goal (having a sane WSGI/HTTP request abstraction), it would be better to only use (and thus depend on) one.

> Regardless, can someone please tell me how to install using Paver?

@lewismc If you look over the `tox.ini` file, it should show what you need to do to install this branch and get it to work. To run the tests, it essentially amounts to:

```bash
pip install -e .[testing,functions,tests]
python setup.py nosetests
```

I just ran that in a fresh install and all 310 tests pass, so I'd be interested to hear whether that works for you. In any case, I'm keen to get a fresh release out so that we can keep Pydap alive, so I'll work on that for the rest of this week.

**Comment by lewismc** [comment]
Cool, I'm testing this branch out as a dependency within [Apache Open Climate Workbench](https://github.com/apache/climate/). I'll continue to update this thread as we continue our tests are well. Thanks @jameshiebert

**Comment by robertodealmeida** [comment]
Regarding Paver, IIRC it was a promising framework at the time and simplified integration with Sphinx and generating the website.

**Comment by lewismc** [comment]
Hi @jameshiebert we'll I've tried this out and it is looking good. I had to manually install the ```coards``` and ```gsw``` dependencies but once that was done via pip all tests pass in Python 3.5.

I also tested this over in Apache OCW and I am experiencing an issue when using Pydap 3.2 installed from this branch. Our test can be found at https://github.com/apache/climate/blob/master/ocw/tests/test_dap.py

```
======================================================================
ERROR: test suite for <class 'ocw.tests.test_dap.TestDap'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/lmcgibbn/miniconda3/lib/python3.5/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/Users/lmcgibbn/miniconda3/lib/python3.5/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/Users/lmcgibbn/miniconda3/lib/python3.5/site-packages/nose/suite.py", line 316, in setupContext
    try_run(context, names)
  File "/Users/lmcgibbn/miniconda3/lib/python3.5/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/usr/local/climate/ocw/tests/test_dap.py", line 32, in setUpClass
    cls.dataset = dap.load(cls.url, 'CGusfc', name=cls.name)
  File "/usr/local/climate/ocw/data_source/dap.py", line 42, in load
    d = open_url(url)
  File "/Users/lmcgibbn/miniconda3/lib/python3.5/site-packages/Pydap-3.2-py3.5.egg/pydap/client.py", line 58, in open_url
    dataset = DAPHandler(url, application).dataset
  File "/Users/lmcgibbn/miniconda3/lib/python3.5/site-packages/Pydap-3.2-py3.5.egg/pydap/handlers/dap.py", line 51, in __init__
    raise_for_status(r)
  File "/Users/lmcgibbn/miniconda3/lib/python3.5/site-packages/Pydap-3.2-py3.5.egg/pydap/net.py", line 21, in raise_for_status
    comment=response.body
webob.exc.HTTPError: 500 Internal Server Error
```

**Comment by jameshiebert** [comment]
Cool, thanks for testing it out @lewismc. I have one final issue on this PR that I'm trying to work out (a single test out of 310 tests failing on python 2). Then I should be good to merge this.

Re: the issue with Apache OCW. Just to clarify, the [failing test case](https://github.com/apache/climate/blob/master/ocw/tests/test_dap.py#L39) isn't actually exercising any Pydap code, is it? It's just a problem with the [coded assumptions about the data's dimensions](https://github.com/apache/climate/blob/master/ocw/data_source/dap.py#L53)?

**Comment by lewismc** [comment]
@jameshiebert 

> Re: the issue with Apache OCW. Just to clarify, the failing test case isn't actually exercising any Pydap code, is it? It's just a problem with the coded assumptions about the data's dimensions?

Correct. As far as I am concerned, your final additions are all that is pending here now.

**Comment by lewismc** [comment]
Fantastic.

**Comment by lewismc** [comment]
@jameshiebert is your intention to merge ```develop``` branch into master?

**Comment by jameshiebert** [comment]
Yeah, definitely. The plan is to merge `develop` into `master` and then do a new release. I was kind of waiting for #11 to get finished up before doing so, but the intent is to do so either way.

**Comment by shoyer** [comment]
Generally folks use `master` for the development version on GitHub, so it probably makes sense to merge the `develop` branch into `master` and ditch `develop`. If necessary, people cut branches (e.g., `3.2.x`) for backporting features to bug-fix releases (but often that's overkill).

**Comment by lewismc** [comment]
@jameshiebert would be great if you could merge into master and then I can take a crack at rebasing the #11 branch off of that.

**Pull request merged**